File: flutter_app/lib/services/preferences_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/accessibility_settings.dart';
import '../models/reminder_preferences.dart';

class PreferencesService {
  static const _settingsKey = 'accessibility_settings';
  static const _reminderPrefsKey = 'reminder_preferences';
  static const _onboardedKey = 'is_onboarded';

  Future<void> saveAccessibilitySettings(AccessibilitySettings settings) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_settingsKey, jsonEncode(settings.toJson()));
  }

  Future<AccessibilitySettings> loadAccessibilitySettings() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_settingsKey);
    if (raw == null) return const AccessibilitySettings();
    return AccessibilitySettings.fromJson(
      jsonDecode(raw) as Map<String, dynamic>,
    );
  }

  Future<void> saveReminderPreferences(ReminderPreferences preferences) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_reminderPrefsKey, jsonEncode(preferences.toJson()));
  }

  Future<ReminderPreferences> loadReminderPreferences() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_reminderPrefsKey);
    if (raw == null) return const ReminderPreferences();
    return ReminderPreferences.fromJson(
      jsonDecode(raw) as Map<String, dynamic>,
    );
  }

  Future<void> saveOnboarded(bool value) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(_onboardedKey, value);
  }

  Future<bool> loadOnboarded() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getBool(_onboardedKey) ?? false;
  }
}

File: react-native-app/src/services/PreferencesService.js
import AsyncStorage from '@react-native-async-storage/async-storage';
import { settingsFromJson } from '../models/AccessibilitySettings';
import { prefsFromJson } from '../models/ReminderPreferences';

const SETTINGS_KEY = 'accessibility_settings';
const REMINDER_PREFS_KEY = 'reminder_preferences';
const ONBOARDED_KEY = 'is_onboarded';

export async function saveAccessibilitySettings(settings) {
  await AsyncStorage.setItem(SETTINGS_KEY, JSON.stringify(settings));
}

export async function loadAccessibilitySettings() {
  try {
    const raw = await AsyncStorage.getItem(SETTINGS_KEY);
    if (!raw) return settingsFromJson({});
    return settingsFromJson(JSON.parse(raw));
  } catch {
    return settingsFromJson({});
  }
}

export async function saveReminderPreferences(prefs) {
  await AsyncStorage.setItem(REMINDER_PREFS_KEY, JSON.stringify(prefs));
}

export async function loadReminderPreferences() {
  try {
    const raw = await AsyncStorage.getItem(REMINDER_PREFS_KEY);
    if (!raw) return prefsFromJson({});
    return prefsFromJson(JSON.parse(raw));
  } catch {
    return prefsFromJson({});
  }
}

export async function saveOnboarded(value) {
  await AsyncStorage.setItem(ONBOARDED_KEY, value ? 'true' : 'false');
}

export async function loadOnboarded() {
  try {
    const raw = await AsyncStorage.getItem(ONBOARDED_KEY);
    return raw === 'true';
  } catch {
    return false;
  }
}
